core/trino-main/src/main/java/io/trino/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -2592,13 +2594,20 @@ private Type coerceToSingleType(StackableAstVisitorContext<Context> context, Str
 
         private void addOrReplaceExpressionCoercion(Expression expression, Type type, Type superType)
         {
-            NodeRef<Expression> ref = NodeRef.of(expression);
-            expressionCoercions.put(ref, superType);
+            addOrReplaceExpressionsCoercion(ImmutableList.of(NodeRef.of(expression)), type, superType);
+        }
+
+        private void addOrReplaceExpressionsCoercion(Iterable<NodeRef<Expression>> expressions, Type type, Type superType)
+        {
+            Map<NodeRef<Expression>, Type> expressionRefTypes = Streams.stream(expressions)
+                    .collect(toImmutableMap(identity(), expression -> superType));
+
+            expressionCoercions.putAll(expressionRefTypes);
             if (typeCoercion.isTypeOnlyCoercion(type, superType)) {
-                typeOnlyCoercions.add(ref);
+                typeOnlyCoercions.addAll(expressionRefTypes.keySet());
```

Comment:
```suggestion
                typeOnlyCoercions.addAll(expressions);
```

Reply:
it would be nice to make it non-Iterable then. otherwise "a materialized copy" (in form of the keyset) is actually a better thing to use

Reply:
It's a collection now